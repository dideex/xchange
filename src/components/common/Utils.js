import React from 'react'

/**
* clean value from the mask, like from 1234 1234 1234 to 123412341234
* @param value{String}, mask{String}
* @return ReactDOM Object{JSX}
* @private
*/
const _clean = (value = '', mask = '') => {
  const maskChars = mask.replace('_', '').split('')
  let raw = value
  for (let i = 0; i < maskChars.length; i++) {
    raw = raw.replace(maskChars[i], '')
  }
  return raw
}

/**
 * Format value from "123456" with mask "__ __ __" to "12 34 56"
 * @param value{String}, mask{string}
 * @return {String}
 * @public
 */
export const format = (value, mask) => {
  if (!mask) return value
  const raw = _clean(value, mask)
  const replsCount = (mask.match(/_/g) || []).length
  let formatted = mask
  const rawLen = raw.length
  for (let i = 0; i < replsCount; i++) {
    if (i >= rawLen) {
      formatted = formatted.substr(0, formatted.lastIndexOf(raw[rawLen - 1]) + 1)
      break
    }
    formatted = formatted.replace('_', raw[i])
  }
  return formatted
}

/**
* Check has string a comma or dot in the end. And return a dot
* @param value{String}
* @return '.'|''{String}
* @private
*/
const _allowLastCharDot = value => {
  const lastChar = value[value.length - 1]
  return lastChar === '.' ? '.' : lastChar === ',' ? '.' : ''
}

/**
* Foramt nubmer to currencie format, from 1123.123123 to '1 123,123'
* @param value{Number}
* @return formattedString{String}
* @public
*/
export const currencyFormat = value => {
  const formatter = new Intl.NumberFormat('ru', 'currency')
  return `${formatter.format(value)}${_allowLastCharDot(`${value}`)}`
}

/**
* Check all object's fields for not been a 'false'
* @param object{Object}
* @return Bool
* @public
*/
export const isAllPropsFalse = object =>
  Object.values(object).filter(argument => argument === false).length ===
  Object.values(object).length


export const ScrollTo = top => {
  const V = 0.3
  const w = window.pageYOffset
  let start = null
  requestAnimationFrame(step)
  function step(time) {
    if (start === null) start = time
    let progress = time - start
    let r =
      top < 0 ? Math.max(w - progress / V, w + top) : Math.min(w + progress / V, w + top)
    window.scrollTo(0, r)
    if (r !== w + top) {
      requestAnimationFrame(step)
    }
  }
}

export const parseOrders = orders =>
  orders &&
  orders.map(
    ({
      _id,
      created,
      inputValue,
      currencyInputLabel,
      outputValue,
      currencyOutputLabel,
      paymentStatus,
      toWallet,
    }) => {
      const date = new Date(Date.parse(created))
      return {
        id: _id,
        paymentStatus,
        toWallet,
        created: `${date.getDate()}.${date.getMonth() + 1}.${date.getFullYear()}`,
        inputValue: `${currencyFormat(inputValue)}, ${currencyInputLabel}`,
        outputValue: `${currencyFormat(outputValue)}, ${currencyOutputLabel}`,
      }
    },
  )

// replace \n to line break in FormattedMessage text
export const linesToParagraphs = (...nodes) =>
  nodes
    .map(node =>
      typeof node === 'string'
        ? node.split('\n').map((text, key) => (
            <span key={key}>
              {text}
              <br />
            </span>
          ))
        : node,
    )
    .reduce((nodes, node) => nodes.concat(node), [])
