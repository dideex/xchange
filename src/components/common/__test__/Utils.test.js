import React from 'react'
import {format, currencyFormat, isAllPropsFalse} from '../Utils'
import {shallow, mount} from 'enzyme'

describe('Utils test', () => {
  it('Format test', () => {
    expect(format('1234','__ __')).toMatchSnapshot()
    expect(format('12341234','__ __')).toMatchSnapshot()
    expect(format('qwe','__ __')).toMatchSnapshot()
    expect(format('1234','_ _ _ _')).toMatchSnapshot()
    expect(format('12341234','_ _ _ _')).toMatchSnapshot()
    expect(format('qwe','_ _ _ _')).toMatchSnapshot()
    expect(format('12345678','____')).toMatchSnapshot()
    expect(format('12345678','_ __ __ __ _')).toMatchSnapshot()
    expect(format('12345678')).toMatchSnapshot()
    expect(format('123')).toMatchSnapshot()
  })

  it('Currency format tests', () => {
    expect(currencyFormat(100)).toMatchSnapshot()
    expect(currencyFormat(1e3)).toMatchSnapshot()
    expect(currencyFormat(1e6)).toMatchSnapshot()
    expect(currencyFormat(1e15)).toMatchSnapshot()
    expect(currencyFormat(.005)).toMatchSnapshot()
    expect(currencyFormat(.100)).toMatchSnapshot()
    expect(currencyFormat(.00)).toMatchSnapshot()
    expect(currencyFormat(100.1)).toMatchSnapshot()
    expect(currencyFormat(100.100000000001)).toMatchSnapshot()
    expect(currencyFormat(-100)).toMatchSnapshot()
    expect(currencyFormat('100')).toMatchSnapshot()
    expect(currencyFormat('100.1')).toMatchSnapshot()
  })

  it('isAllPropsFalse tests', () => {
    expect(isAllPropsFalse({a: 1, b: 2, c: 3})).toBeFalsy()
    expect(isAllPropsFalse({a: 1, b: 2, c: null})).toBeFalsy()
    expect(isAllPropsFalse({a: false, b: false, c: false})).toBeTruthy()
  })
})