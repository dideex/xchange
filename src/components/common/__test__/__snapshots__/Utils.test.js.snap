// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils test Currency format tests 1`] = `"100"`;

exports[`Utils test Currency format tests 2`] = `"1,000"`;

exports[`Utils test Currency format tests 3`] = `"1,000,000"`;

exports[`Utils test Currency format tests 4`] = `"1,000,000,000,000,000"`;

exports[`Utils test Currency format tests 5`] = `"0.005"`;

exports[`Utils test Currency format tests 6`] = `"0.1"`;

exports[`Utils test Currency format tests 7`] = `"0"`;

exports[`Utils test Currency format tests 8`] = `"100.1"`;

exports[`Utils test Currency format tests 9`] = `"100.1"`;

exports[`Utils test Currency format tests 10`] = `"-100"`;

exports[`Utils test Currency format tests 11`] = `"100"`;

exports[`Utils test Currency format tests 12`] = `"100.1"`;

exports[`Utils test Format test 1`] = `"12 34"`;

exports[`Utils test Format test 2`] = `"12 34"`;

exports[`Utils test Format test 3`] = `"qw e"`;

exports[`Utils test Format test 4`] = `"1 2 3 4"`;

exports[`Utils test Format test 5`] = `"1 2 3 4"`;

exports[`Utils test Format test 6`] = `"q w e"`;

exports[`Utils test Format test 7`] = `"1234"`;

exports[`Utils test Format test 8`] = `"1 23 45 67 8"`;

exports[`Utils test Format test 9`] = `"12345678"`;

exports[`Utils test Format test 10`] = `"123"`;
